src/core/memory.cc

```diff
@@ -245,6 +246,76 @@ static uint64_t get(types stat_type)
 
 }
 
+// Taken from
+// https://cs.android.com/android/platform/superproject/+/master:external/perfetto/src/profiling/memory/sampler.h
+// See: https://perfetto.dev/docs/design-docs/heapprofd-sampling for the background
+//
+// Note there is a few differences for how we use this. Right now we don't
+// respect the samplesize/weight that we return here. The reason is that that
+// would require storing the weight in addition to the alloation site ptr as on
+// free we need to know how much a sample accounted for. Hence, for now we
+// simple always use the sample frequency.
+class Sampler {
```

Comment:
> The original source code is under an Apache 2.0 license. Does this license allow for copying subsets of the work with attribution?

I am fairly certain but will double check.

 > We have a goal of up-streaming this PR to scylladb/seastar. Do they accept code under an Apache 2.0 license? And if so do they accept code that isn't formatted according to their styling guide?

I was worried at first as well but then saw that seastar is actually also apache2 licensed so I hope this shouldn't be a problem.

Re formatting, I just dumped the whole thing in for now but I will probably take it apart a bit so that we can safe a couple instructions at which point the code will probably not really resemble the original anymore and I will be able to format as per seastar standard (not saying I would want to remove the mention of the original one).